Sources/OpenCombine/Publishers/Publishers.ReplaceError.swift

```diff
@@ -0,0 +1,134 @@
+//
+//  Publishers.ReplaceError.swift
+//  OpenCombine
+//
+//  Created by Bogdan Vlad on 8/29/19.
+//
+
+extension Publishers {
+    /// A publisher that replaces any errors in the stream with a provided element.
+    public struct ReplaceError<Upstream>: Publisher where Upstream: Publisher {
+
+        /// The kind of values published by this publisher.
+        public typealias Output = Upstream.Output
+
+        /// The kind of errors this publisher might publish.
+        ///
+        /// Use `Never` if this `Publisher` does not publish errors.
+        public typealias Failure = Never
+
+        /// The element with which to replace errors from the upstream publisher.
+        public let output: Upstream.Output
+
+        /// The publisher from which this publisher receives elements.
+        public let upstream: Upstream
+
+        public init(upstream: Upstream,
+                    output: Publishers.ReplaceError<Upstream>.Output) {
+            self.upstream = upstream
+            self.output = output
+        }
+
+        /// This function is called to attach the specified `Subscriber`
+        /// to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<SubscriberType: Subscriber>(subscriber: SubscriberType)
+            where Upstream.Output == SubscriberType.Input,
+                  SubscriberType.Failure == Failure
+        {
+            let replaceErrorSubscriber = _ReplaceError<Upstream, SubscriberType>(
+                downstream: subscriber,
+                output: output
+            )
+            upstream.subscribe(replaceErrorSubscriber)
+        }
+    }
+}
+
+extension Publisher {
+    /// Replaces any errors in the stream with the provided element.
+    ///
+    /// If the upstream publisher fails with an error, this publisher emits the provided
+    /// element, then finishes normally.
+    /// - Parameter output: An element to emit when the upstream publisher fails.
+    /// - Returns: A publisher that replaces an error from the upstream publisher with
+    ///            the provided output element.
+    public func replaceError(with output: Self.Output) -> Publishers.ReplaceError<Self> {
+        return Publishers.ReplaceError(upstream: self, output: output)
+    }
+}
+
+private final class _ReplaceError<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscriber,
+      CustomStringConvertible,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Failure = Upstream.Failure
+
+    var hasAnyDownstreamDemand: Bool = false
+    var hasFailed: Bool = false
+    var description: String { return "ReplaceError" }
+
+    private let _lock = Lock(recursive: true)
```

Comment:
Let's not do locking for now. Apple's Combine uses locks very selectively.

I plan to audit the code for thread safety at some point in the future, but right now after reverse engineering some of the Combine internals I can tell that most operators don't do locking at all.